### ISSUE TITLE
Poor Mock Server Performance on Windows

### ISSUE THREAD

**Issue by adamrodger** [pact-foundation/pact-reference#147]
See [related issue in PactNet](https://github.com/pact-foundation/pact-net/issues/323) for lots more details, but the tl;dr is the mock server takes ~2s per request on Windows, but this doesn't happen on Linux.

This means that a suite of consumer tests can take minutes to execute in PactNet, which is much worse performance than the older version using the Ruby backend.

I'm not sure why it happens, but I have confirmed that it's definitely caused by the Pact Mock Server by switching out the calls to use a simple echo server running in Docker, and that fixes the performance issues immediately. Again, further details on the related ticket.

**Comment by adamrodger** [comment]
Here's the mock server logs, but unfortunately they'd don't have timestamps on so if the problem is happening somewhere after the request has been received then it's not really possible to tell where just from this:

```
[DEBUG][pact_mock_server::hyper_server] Creating pact request from hyper request
[DEBUG][pact_mock_server::hyper_server] Extracting query from uri /events
[INFO][pact_mock_server::hyper_server] Received request Request ( method: GET, path: /events, query: None, headers: Some({"host": ["localhost:63227"], "accept": ["application/json"], "authorization": ["Bearer SomeValidAuthToken"]}), body: Empty )
[INFO][pact_matching] comparing to expected Request ( method: GET, path: /events, query: None, headers: Some({"Accept": ["application/json"], "Authorization": ["Bearer SomeValidAuthToken"]}), body: Missing )
[DEBUG][pact_matching]      body: ''
[DEBUG][pact_matching]      matching_rules: MatchingRules { rules: {HEADER: MatchingRuleCategory { name: HEADER, rules: {} }, PATH: MatchingRuleCategory { name: PATH, rules: {} }} }
[DEBUG][pact_matching]      generators: Generators { categories: {} }
[DEBUG][pact_matching::matchers] String -> String: comparing '/events' to '/events' using Equality
[DEBUG][pact_matching] expected content type = '*/*', actual content type = '*/*'
[DEBUG][pact_matching] content type header matcher = 'None'
[DEBUG][pact_matching::matchers] String -> String: comparing 'Bearer SomeValidAuthToken' to 'Bearer SomeValidAuthToken' using Equality
[DEBUG][pact_matching] --> Mismatches: []
[DEBUG][pact_mock_server::hyper_server] Test context = {"mockServer": Object({"href": String("http://127.0.0.1:63227"), "port": Number(63227)})}
[INFO][pact_mock_server::hyper_server] Request matched, sending response Response ( status: 200, headers: Some({"Content-Type": ["application/json; charset=utf-8"]}), body: Present(356 bytes) )
[DEBUG][pact_mock_server::hyper_server]      body: '[{"eventId":"45d80d13-d5a2-48d7-8353-cbb4c0eaabf5","eventType":"SearchView","timestamp":"2014-06-30T01:37:41.0660548"},{"eventId":"83f9262f-28f1-4703-ab1a-8cfd9e8249c9","eventType":"DetailsView","timestamp":"2014-06-30T01:37:52.2618864"},{"eventId":"3e83a96b-2a0c-49b1-9959-26df23f83aeb","eventType":"SearchView","timestamp":"2014-06-30T01:38:00.8518952"}]'
```

**Comment by adamrodger** [comment]
But weirdly if I pause my test in the middle and manually call the mock server with `curl` then the request is reasonably fast:

```bash
$ time curl --silent -H "Accept: application/json" -H "Authorization: Bearer SomeValidAuthToken" http://localhost:63274/events
[{"eventId":"45d80d13-d5a2-48d7-8353-cbb4c0eaabf5","eventType":"SearchView","timestamp":"2014-06-30T01:37:41.0660548"},{"eventId":"83f9262f-28f1-4703-ab1a-8cfd9e8249c9","eventType":"DetailsView","timestamp":"2014-06-30T01:37:52.2618864"},{"eventId":"3e83a96b-2a0c-49b1-9959-26df23f83aeb","eventType":"SearchView","timestamp":"2014-06-30T01:38:00.8518952"}]
real    0m0.250s
user    0m0.000s
sys     0m0.015s
```

Could it be that the server isn't closing the connection properly or something and so the client sits and waits for 2 seconds before hanging up? Something very weird is definitely going on anyway given it only happens on Windows

**Comment by adamrodger** [comment]
Important update from the attached ticket:

If you use the standard approach (where the `baseUri` is `http://localhost:<port>` of the mock server) then the tests are really slow:

![image](https://user-images.githubusercontent.com/508586/133940161-7962e734-104e-4391-85f8-3fceafa06d17.png)
![image](https://user-images.githubusercontent.com/508586/133940175-aaf1dafc-e15a-4f8f-afa0-8008302a4f0e.png)

BUT if you replace `localhost` with `127.0.0.1` then the tests are fast:

![image](https://user-images.githubusercontent.com/508586/133940124-d9f50826-c071-4bb7-9d09-77f38db56e33.png)
![image](https://user-images.githubusercontent.com/508586/133940141-9cc25b27-4fc6-492d-94ab-2dbbdc30e3df.png)

I can't really explain why that happens at the moment? Perhaps the `hyper` server has some issue or different behaviour on Windows when you use `localhost` instead of `127.0.0.1`?

**Comment by adamrodger** [comment]
Erm, well this code looks rather problematic and I have no idea how it's even working:

![image](https://user-images.githubusercontent.com/508586/133940535-d5af27f2-f70f-4b42-9f36-074e4ef11839.png)

`this.host` is set to `IpAddress.Loopback`, which is just an enum defined in PactNet with nothing special about it at all. As far as I can tell that means the URL should (incorrectly) be set to `http://Loopback:<port>` but instead it gets set to `http://localhost:<port>`.

This is likely because the `IpAddress.Loopback.ToString() == "0"` and `http://0:<port>` is invalid, and thus `new Uri("http://0:<port>)` just defaults to `localhost` when it fails to parse 😱 

This is a typo in the original though - it should be using the IP address which is resolved further up so that the address resolves to `http://127.0.0.1:<port>` instead, which also works around the performance issue.
